kart/sqlalchemy_import_source.py

```diff
@@ -0,0 +1,242 @@
+import functools
+import os
+import sys
+
+import click
+import sqlalchemy
+
+from .exceptions import (
+    NotFound,
+    NotYetImplemented,
+    NO_TABLE,
+)
+from .geometry import Geometry
+from .import_source import ImportSource
+from .sqlalchemy import DbType, separate_last_path_part
+from .output_util import dump_json_output
+from .utils import ungenerator, chunk
+
+
+class SqlAlchemyImportSource(ImportSource):
+    """
+    ImportSource that uses SqlAlchemy directly to import into Kart.
+    Currently only GPKG is supported, but in theory should work for Postgis, SQL Server, MySQL.
+    """
+
+    CURSOR_SIZE = 10000
+
+    @classmethod
+    def open(cls, spec):
+        db_type = DbType.from_spec(spec)
+        if db_type is None:
+            raise cls._bad_import_source_spec(spec)
+
+        # TODO - add support for other DB types.
+        if db_type is not DbType.GPKG:
+            raise NotYetImplemented(
+                "Only GPKG is currently supported by the SqlAlchemyImportSource"
+            )
+
+        path_length = db_type.path_length(spec)
+        longest_allowed_path_length = db_type.path_length_for_table
+        shortest_allowed_path_length = max(
+            db_type.path_length_for_table_container - 1, 0
+        )
+
+        if (
+            path_length > longest_allowed_path_length
+            or path_length < shortest_allowed_path_length
+        ):
```

Comment:
```suggestion
        if not (
            shortest_allowed_path_length <= path_length <= longest_allowed_path_length
        ):
```

Reply:
Ooh I think this is the first language I've used that can do this